Add a blog post demoing image similarity system with 🤗 Transformers and Datasets (#663)
huggingface/blog: image-similarity.md

---

> title: Building an Image Similarity System with 🤗 Datasets and Transformers

Reviewer: Not sure if better for SEO to have the actual words "Hugging Face" here in addition to the emoji (@osanseviero do you know?)

Reviewer 2: I would use actual words here :)

Author: To keep the title relatively shorter, how about "Image Similarity System with Hugging Face Datasets and Transformers"?

Next, for convenience, we create a dictionary mapping the identifiers of the candidate images and their embeddings we just computed. This dictionary will come in handy when performing retrievals.

```py
id_embeddings = {}

for id in tqdm(range(len(candidate_subset_emb))):
    embeddings, label = (
        candidate_subset_emb[id]["embeddings"],
        candidate_subset_emb[id]["labels"],
    )

    # Create a unique indentifier.
    entry = str(id) + "_" + str(label)

    id_embeddings.update({entry: embeddings})
```

We can now code up another utility to fetch the candidates images matching a query image:

```py
def fetch_similar(image: PIL.Image, top_k: int = 5):
    """Fetches the `top_k` similar images with `image` as the query."""
    similarity_mapping = {}

    image_transformed = transformation_chain(image).unsqueeze(0)
    new_batch = {"pixel_values": image_transformed.to(device)}

    with torch.no_grad():
        query_embeddings = model(**new_batch).last_hidden_state[:, 0].cpu()

    for id_entry in id_embeddings:
        candidate_embeddings = torch.tensor(id_embeddings[id_entry]).unsqueeze(0)
        sim_score = compute_scores(candidate_embeddings, query_embeddings)

>         similarity_mapping.update({id_entry: sim_score})

Reviewer 3: I think we can recommend using a matrix and and call `compute_scores` only once here no ? The for loop makes querying slower

Author: Could you elaborate? Yes, for loop querying is definitely slower. To mitigate that problem I mentioned the use of FAISS and other things.

Reviewer 3: For example you can pass the full matrix of embeddings to cosine_similarity()

Author: Raised a PR here: https://github.com/huggingface/notebooks/pull/291. Once that is merged, I will make the necessary changes and let you and the other reviewers know.

Author: @lhoestq I have changed this portion to enable faster querying. 

@nateraw would you be able to take a final pass on the changes? The latest changes basically reflect modifications from https://github.com/huggingface/notebooks/pull/291.

```py
dataset_with_embeddings.add_faiss_index(column="embeddings")
```

Once the index is built, `dataset_with_embeddings` can be used to retrieve the nearest examples given query embeddings with [`get_nearest_examples()`](https://huggingface.co/docs/datasets/v2.7.1/en/package_reference/main_classes#datasets.Dataset.get_nearest_examples):

```py
scores, retrieved_examples = dataset_with_embeddings.get_nearest_examples("embeddings", qi_embedding, k=top_k)
```

The method returns scores and corresponding candidate examples. To know more, you can check out the [official documentation](https://huggingface.co/docs/datasets/faiss_es) and [this notebook](https://colab.research.google.com/gist/sayakpaul/5b5b5a9deabd3c5d8cb5ef8c7b4bb536/image_similarity_faiss.ipynb).

Finally, you can try out the following Space that builds a mini image similarity application:

<div class="hidden xl:block">
<div style="display: flex; flex-direction: column; align-items: center;">
<iframe
	src="https://sayakpaul-fetch-similar-images.hf.space"
	frameborder="0"
	width="850"
	height="450"

>     allow="accelerometer; ambient-light-sensor; autoplay; battery; camera; document-domain; encrypted-media; fullscreen; geolocation; gyroscope; layout-animations; legacy-image-formats; magnetometer; microphone; midi; oversized-images; payment; picture-in-picture; publickey-credentials-get; sync-xhr; usb; vr ; wake-lock; xr-spatial-tracking"

Reviewer: Seeing this in Chrome dev tools

<img width="870" alt="Screen Shot 2023-01-03 at 9 00 24 PM" src="https://user-images.githubusercontent.com/32437151/210470967-2a1f36d9-aabf-4107-8d26-41fe93e16579.png">

Author: Not sure about this. I took it from [here](https://raw.githubusercontent.com/huggingface/blog/main/tf-xla-generate.md).

Author: I checked the preview and it seemed to work fine with the `iframe`.

Author: @osanseviero TIL!

I incorporated it but the Space has an issue reported [here](https://huggingface.slack.com/archives/C034N0A7H09/p1672923569108359).

Reviewer 2: Thanks a lot!

Author: Rendering beautifully in the preview too:

<img width="935" alt="image" src="https://user-images.githubusercontent.com/22957388/210829432-cd1d6e01-b09e-48d1-b676-1c233f7ba9ea.png">

Although the post uses a ViT-based model ([`nateraw/vit-base-beans`](https://huggingface.co/nateraw/vit-base-beans)) and a particular dataset ([Beans](https://huggingface.co/datasets/beans)), it can be extended to use other models supporting vision modality and other image datasets. Some notable models, you could try:

* [Swin Transformer](https://huggingface.co/docs/transformers/model_doc/swin)
* [ConvNeXT](https://huggingface.co/docs/transformers/model_doc/convnext)
* [RegNet](https://huggingface.co/docs/transformers/model_doc/regnet)

Also, the approach presented in the post can potentially be extended to other modalities as well.

To study the fully working image-similarity system, you can refer to the Colab Notebook linked at the beginning. 

## How do we define similarity?

To build this system, we first need to define how we want to compute the similarity between two images. One widely popular practice is to compute dense representations (embeddings) of the given images and then use the [cosine similarity metric](https://en.wikipedia.org/wiki/Cosine_similarity) to determine how similar the two images are. 

> For this post, we'll be using “embeddings” to represent images in vector space. This gives us a nice way to meaningfully compress the high-dimensional pixel space of images (224 x 224 x 3, for example) to something much lower dimensional (768, for example). The primary advantage of doing this is the reduced computation time in the subsequent steps.

Reviewer 2: ```suggestion
For this post, we'll use “embeddings” to represent images in vector space. This gives us a nice way to meaningfully compress the high-dimensional pixel space of images (224 x 224 x 3, for example) to something much lower dimensional (768, for example). The primary advantage of doing this is the reduced computation time in the subsequent steps.
```

<div align="center">
<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/image_similarity/embeddings.png" width=700/>
</div>

## Computing embeddings

> To compute the embeddings from the images, we'll use a vision model that has some understanding of how to represent the input images in the vector space. This type of models is also commonly referred to as image encoders.
> 
> For loading the model, we leverage the [`AutoModel` class](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModel). It provides an interface for us to load any compatible model checkpoint from the 🤗 Hub. Alongside the model, we also load the processor associated with the model for data preprocessing. 

Reviewer 2: ```suggestion
To compute the embeddings from the images, we'll use a vision model that has some understanding of how to represent the input images in the vector space. This type of model is also commonly referred to as image encoder.

For loading the model, we leverage the [`AutoModel` class](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModel). It provides an interface for us to load any compatible model checkpoint from the Hugging Face Hub. Alongside the model, we also load the processor associated with the model for data preprocessing. 
```

```py
from transformers import AutoFeatureExtractor, AutoModel

model_ckpt = "nateraw/vit-base-beans"
extractor = AutoFeatureExtractor.from_pretrained(model_ckpt)
model = AutoModel.from_pretrained(model_ckpt)
```

In this case, the checkpoint was obtained by fine-tuning a [Vision Transformer based model](https://huggingface.co/google/vit-base-patch16-224-in21k) on the [`beans` dataset](https://huggingface.co/datasets/beans).

Some questions that might arise here:

**Q1**: Why did we not use `AutoModelForImageClassification`?

This is because we want to obtain dense representations of the images and not discrete categories, which are what `AutoModelForImageClassification` would have provided.

**Q2**: Why this checkpoint in particular?

> We're using a specific dataset to build the system as mentioned earlier. So, instead of using a generalist model (like the [ones trained on the ImageNet-1k dataset](https://huggingface.co/models?dataset=dataset:imagenet-1k&sort=downloads), for example), it's better to use a model that has been fine-tuned on the dataset being used. That way, the underlying model has a better understanding of the input images.

Reviewer 2: ```suggestion
As mentioned earlier, we're using a specific dataset to build the system. So, instead of using a generalist model (like the [ones trained on the ImageNet-1k dataset](https://huggingface.co/models?dataset=dataset:imagenet-1k&sort=downloads), for example), it's better to use a model that has been fine-tuned on the dataset being used. That way, the underlying model better understands the input images.
```

**Note** that you can also use a checkpoint that was obtained through self-supervised pre-training. The checkpoint doesn't necessarily have to come from supervised learning. In fact, if pre-trained well, self-supervised models can [yield](https://ai.facebook.com/blog/dino-paws-computer-vision-with-self-supervised-transformers-and-10x-more-efficient-training/) impressive retrieval
performance. 

Now that we have a model for computing the embeddings, we need some candidate images to query against. 

## Loading a dataset for candidate images

In some time, we'll be building hash tables mapping the candidate images to hashes. During the query time, we'll use these hash tables. We'll talk more about hash tables in the respective section but for now, to have a set of candidate images, we will use the `train` split of the [`beans` dataset](https://huggingface.co/datasets/beans). 

```py
from datasets import load_dataset

dataset = load_dataset("beans")
```

This is how a single sample from the training split looks like:

> <div align="center">

Reviewer 2: Maybe also show the code to generate this

Author: I avoided it intentionally to reduce the lines of code from the blog.

Reviewer 2: I feel a good % of users will just go through the blog post copy pasting the code cells, so it's good to be complete (at least to the level that they can have everything working by running all the code in the blog post). I think showing code that uses functions  that are not mentioned nor defined in the blog are more likely to increase confusion than help

Author: Understood. 

But this is not the case here no? There's no undefined variable or function in this particular section, hence my comment.

Reviewer 2: Fair for this section. It was more of a general comment though, there is `transformation_chain` for example.

Author: I have explained it in the code comments:)

    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/image_similarity/beans.png" width=600/>
</div>

> The dataset has got three columns / features:

Reviewer 2: ```suggestion
The dataset has three features:
```

```bash

> {'image_file_path': Value(dtype='string', id=None),

Reviewer 2: Maybe add `dataset["train"].features` so people know how you got these

 'image': Image(decode=True, id=None),
 'labels': ClassLabel(names=['angular_leaf_spot', 'bean_rust', 'healthy'], id=None)}
```

## The process of finding similar images

Below, you can find a pictorial overview of the process underlying fetching similar images. 

<div align="center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/image_similarity/fetch-similar-process.png">
</div>

Breaking down the above figure a bit, we have:

> 1. Extract the embeddings from the candidate images (`candidate_subset`) storing them in a matrix.

Reviewer 2: ```suggestion
1. Extract the embeddings from the candidate images (`candidate_subset`), storing them in a matrix.
```

2. Take a query image and extract its embeddings. 
3. Iterate over the embedding matrix (computed in step 1) and compute the similarity score between the query embedding and the current candidate embeddings. We usually maintain a dictionary-like mapping  maintaining a correspondence between some identifier of the candidate image and the similarity scores. 

> 4. Sort the mapping structure w.r.t the similarity scores and return the identifiers underlying. We use these identifiers to fetch the candidate samples.

Reviewer 2: ```suggestion
4. Sort the mapping structure w.r.t the similarity scores and return the underlying identifiers. We use these identifiers to fetch the candidate samples.
```

We can write a simple utility and `map()` it to our dataset of candidate images to compute the embeddings efficiently. 

```py

> def extract_embeddings(model: torch.nn.Module):

Reviewer 2: Maybe add an `import torch` here

    """Utility to compute embeddings."""
    device = model.device

    def pp(batch):
        images = batch["image"]
        image_batch_transformed = torch.stack(

>             [transformation_chain(image) for image in images]

Reviewer 2: `transformation_chain` is not defined or explained here. Maybe I would add some explanation about it

Author: Added in comments.

        )
        new_batch = {"pixel_values": image_batch_transformed.to(device)}
        with torch.no_grad():
            embeddings = model(**new_batch).last_hidden_state[:, 0].cpu()
        return {"embeddings": embeddings}

    return pp
```

And we can map `extract_embeddings()` like so:

```py
extract_fn = extract_embeddings(model.to(device))

> candidate_subset_emb = candidate_subset.map(extract_fn, batched=True, batch_size=batch_size)

Reviewer 2: You should also add the code to have the `candidate_subset` and `device`

Author: Done.

    ids = list(map(lambda x: int(x.split("_")[0]), id_entries))
    labels = list(map(lambda x: int(x.split("_")[-1]), id_entries))
    return ids, labels
```

## Perform a query

Given all the utilities, we're equipped to do a similarity search. Let's have a query image from the `test` split of
the `beans` dataset:

```py
test_idx = np.random.choice(len(dataset["test"]))
test_sample = dataset["test"][test_idx]["image"]
test_label = dataset["test"][test_idx]["labels"]

sim_ids, sim_labels = fetch_similar(test_sample)
print(f"Query label: {test_label}")
print(f"Top 5 candidate labels: {sim_labels}")
```

Leads to:

```
Query label: 0
Top 5 candidate labels: [0, 0, 0, 0, 0]
```

Seems like our system got the right set of similar images. When visualized, we'd get:

<div align="center">
<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/image_similarity/results_one.png">
</div>

## Further extensions and conclusions

We now have a working image similarity system. But in reality, you'll be dealing with a lot more candidate images. Taking that into consideration, our current procedure has got multiple drawbacks:

> * If we store the embeddings as is then the memory requirements can shoot up pretty quickly especially when dealing with millions of candidate images. The embeddings are 768-d in our case, which can be still quite high in the large-scale regime. 

Reviewer 2: ```suggestion
* If we store the embeddings as is, the memory requirements can shoot up quickly, especially when dealing with millions of candidate images. The embeddings are 768-d in our case, which can still be relatively high in the large-scale regime. 
```

* Having high-dimensional embeddings have a direct effect on the subsequent computations involved in the retrieval part. 

> So, if we can somehow reduce the dimensionality of the embeddings without disturbing their meaning, we can still maintain a good trade-off between speed and retrieval quality. The [accompanying Colab Notebook](https://colab.research.google.com/github/huggingface/notebooks/blob/main/examples/image_similarity.ipynb) of this post implements and demonstrates utilities for achieving this with random projection and locality-sensitive hashing. 

Reviewer 2: ```suggestion
If we can somehow reduce the dimensionality of the embeddings without disturbing their meaning, we can still maintain a good trade-off between speed and retrieval quality. The [accompanying Colab Notebook](https://colab.research.google.com/github/huggingface/notebooks/blob/main/examples/image_similarity.ipynb) of this post implements and demonstrates utilities for achieving this with random projection and locality-sensitive hashing. 
```

    "embeddings", qi_embedding, k=top_k
)
```

The method returns scores and corresponding candidate examples. To know more, you can check out the [official documentation](https://huggingface.co/docs/datasets/faiss_es) and [this notebook](https://colab.research.google.com/gist/sayakpaul/5b5b5a9deabd3c5d8cb5ef8c7b4bb536/image_similarity_faiss.ipynb).

Finally, you can try out the following Space that builds a mini image similarity application:

<div class="hidden xl:block">
<div style="display: flex; flex-direction: column; align-items: center;">
<iframe
	src="https://sayakpaul-fetch-similar-images.hf.space"
	frameborder="0"
	width="850"
	height="450"
    allow="accelerometer; ambient-light-sensor; autoplay; battery; camera; document-domain; encrypted-media; fullscreen; geolocation; gyroscope; layout-animations; legacy-image-formats; magnetometer; microphone; midi; oversized-images; payment; picture-in-picture; publickey-credentials-get; sync-xhr; usb; vr ; wake-lock; xr-spatial-tracking"
    sandbox="allow-forms allow-modals allow-popups allow-popups-to-escape-sandbox allow-same-origin allow-scripts allow-downloads"
></iframe>
</div>
</div>

> WIn this post, we ran through a quickstart for building image similarity systems. If you found this post interesting, we highly recommend building on top of the concepts we discussed here so you can get more comfortable with the inner workings.

Reviewer 3: ```suggestion
In this post, we ran through a quickstart for building image similarity systems. If you found this post interesting, we highly recommend building on top of the concepts we discussed here so you can get more comfortable with the inner workings.
```

<div align="center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/image_similarity/fetch-similar-process.png">
</div>

Breaking down the above figure a bit, we have:

1. Extract the embeddings from the candidate images (`candidate_subset`), storing them in a matrix.
2. Take a query image and extract its embeddings. 
3. Iterate over the embedding matrix (computed in step 1) and compute the similarity score between the query embedding and the current candidate embeddings. We usually maintain a dictionary-like mapping  maintaining a correspondence between some identifier of the candidate image and the similarity scores. 
4. Sort the mapping structure w.r.t the similarity scores and return the underlying identifiers. We use these identifiers to fetch the candidate samples.

We can write a simple utility and `map()` it to our dataset of candidate images to compute the embeddings efficiently. 

```py
import torch 

def extract_embeddings(model: torch.nn.Module):
    """Utility to compute embeddings."""
    device = model.device

    def pp(batch):
        images = batch["image"]

>         # `transformation_chain` is a compostion of preprocessing

Author: @osanseviero here.

Now that the LSH tables have been built, we can use them to query them with images. 

## Inference

In this secton, we'll take query images from the `test` split of our dataset and retrieve the similar images from the set of candidate images we have. 

```py
def visualize_lsh(lsh_class: BuildLSHTable, top_k: int = 5):
    idx = np.random.choice(len(dataset["test"]))

    image = dataset["test"][idx]["image"]
    label = dataset["test"][idx]["labels"]
    results = lsh_class.query(image)

    candidates = []
    labels = []
    overlaps = []

    for idx, r in enumerate(sorted(results, key=results.get, reverse=True)):
        if idx == (top_k - 1):
            break
        image_id, label = r.split("_")[0], r.split("_")[1]
        candidates.append(dataset["train"][int(image_id)]["image"])
        labels.append(label)
        overlaps.append(results[r])

    candidates.insert(0, image)
    labels.insert(0, label)

    plot_images(candidates, labels)
```

Let's now put `visualize_lsh()` to use:

```py
visualize_lsh(lsh_builder)
```

<div align="center">
    <img src="https://i.ibb.co/L5rsj3r/image.png"/> <!-- TODO -->
</div>

## Conclusion

> TODO

Author: I want to add some call for actions here, including the point I made [here](https://github.com/huggingface/notebooks/pull/270#issuecomment-1321503361) regarding ONNX to optimize the runtime. But ONNX is acting crazing. 

But will add other things too.
